src/lib/replication/repl_dev/raft_state_machine.cpp

```diff
@@ -297,14 +297,22 @@ void RaftStateMachine::create_snapshot(nuraft::snapshot& s, nuraft::async_result
 
 int RaftStateMachine::read_logical_snp_obj(nuraft::snapshot& s, void*& user_ctx, ulong obj_id, raft_buf_ptr_t& data_out,
                                            bool& is_last_obj) {
+    // For Nuraft baseline resync, we separate the process into two layers: HomeStore layer and Application layer.
+    // We use the highest bit of the obj_id to indicate the message type: 0 is for HS, 1 is for Application.
+    if ((obj_id & snp_obj_id_type_mask) == 0) {
+        // This is the preserved msg for homestore to resync data
+        m_rd.create_snp_resync_data(data_out);
+        is_last_obj = false;
+        return 0;
+    }
     auto snp_ctx = std::make_shared< nuraft_snapshot_context >(s);
-    auto snp_data = std::make_shared< snapshot_data >();
+    auto snp_data = std::make_shared< snapshot_obj >();
     snp_data->user_ctx = user_ctx;
     snp_data->offset = obj_id;
     snp_data->is_last_obj = is_last_obj;
 
     // Listener will read the snapshot data and we pass through the same.
-    int ret = m_rd.m_listener->read_snapshot_data(snp_ctx, snp_data);
+    int ret = m_rd.m_listener->read_snapshot_obj(snp_ctx, snp_data);
     if (ret < 0) return ret;
```

Comment:
Is anyone catching this error code?

Reply:
NuRaft will handle this error and finish the snapshot. In HomeObject read logic, we add some retries for get_blob.
https://github.com/eBay/NuRaft/blob/cda7dffee79bc0899d51917a9765eb0392ae958f/src/handle_snapshot_sync.cxx#L207